webui/react/src/components/NotebookModal.tsx

```diff
@@ -0,0 +1,309 @@
+}
+
+interface ResourceInfo {
+  hasCPU: boolean;
+  hasGPU: boolean;
+  showResourceType: boolean;
+}
+
+const NotebookModal: React.FC<NotebookModalProps> = (
+  { visible = false, onLaunch, ...props }: NotebookModalProps,
+) => {
+
+  const [ showFullConfig, setShowFullConfig ] = useState(false);
+  const [ fields, dispatch ] = useNotebookForm();
+  const [ config, setConfig ] = useState<RawJson>({});
+
+  const fetchConfig = useCallback(async () => {
+    try {
+      const config = await previewNotebook(
+        fields.slots,
+        fields.template,
+        fields.name,
+        fields.pool,
+      );
+      setConfig({ config: yaml.dump(config) });
+
+    } catch {}
+  }, [ fields ]);
+
+  useEffect(() => {
+    if (showFullConfig) fetchConfig();
+  }, [ showFullConfig, fetchConfig ]);
+
+  const handleSecondary = useCallback(() => {
+    setShowFullConfig(show => !show);
+  }, [ ]);
+
+  const handleCreateEnvironment = useCallback(
+    () => {
+      if (showFullConfig) {
+        launchNotebook(config);
+      } else {
+        launchNotebook(
+          undefined,
+          fields.slots,
+          fields.template,
+          fields.name,
+          fields.pool,
+        );
+      }
+      if (onLaunch) onLaunch();
+    },
+    [ showFullConfig, onLaunch, fields, config ],
+  );
+
+  return <Modal
+    footer={<>
+      <Button onClick={handleSecondary}>{showFullConfig ? 'Edit Form' : 'Edit Full Config'}</Button>
+      <Button
+        type="primary"
+        onClick={handleCreateEnvironment}>Launch</Button>
+    </>}
+    title="Launch JupyterLab"
+    visible={visible}
+    width={540}
+    {...props}>
+    {showFullConfig ?
+      <NotebookFullConfig config={config} onChange={(newConfig) => setConfig(newConfig)} /> :
+      <NotebookForm fields={fields} onChange={dispatch} />
+    }
+  </Modal>;
+};
+
+const NotebookFullConfig:React.FC<FullConfigProps> = (
+  { config, onChange }:FullConfigProps,
+) => {
+  const [ field, setField ] = useState([ { name: 'config', value: '' } ]);
+
+  useEffect(() => {
+    setField([ { name: 'config', value: yaml.dump(config) } ]);
+  }, [ config ]);
+
+  const extractConfig = useCallback((fieldData) => {
+    onChange(JSON.parse(fieldData[0].value));
+  }, [ onChange ]);
+
+  return <Form
+    fields={field}
+    onFieldsChange={(_, allFields) => {
+      extractConfig(allFields);
+    }}>
+    <div className={css.note}>
+      <Link external path="/docs/reference/command-notebook-config.html">
+    Read about notebook settings
+      </Link>
+    </div>
+    <React.Suspense fallback={<div className={css.loading}><Spinner /></div>}>
+      <Item
+        name="config"
+        noStyle
+        rules={[ { message: 'Invalid YAML', required: true }, () => ({
+          validator(_, value) {
+            try {
+              yaml.load(value);
+              return Promise.resolve();
+            } catch(err) {
+              return Promise.reject(new Error('Invalid YAML'));
```

Comment:
**question:** does `err` hold any info on where the parsing error might be in the YAML config? It might be better to show that error if provides the user clues as to how to fix the YAML as these config files can get large for certain users.